### PR TITLE
Add Mo-Fr shortcut to weekday selection

### LINKED ISSUE #115: Jours d'ouverture : raccourcis (Lun-Ven, Lun-Sam, Lun-Dim)
Beaucoup de magasins ont le même horaire pour plusieurs jours. Jusqu'ici, il faut cocher tout les jours l'un après l'autre. Il serait donc util de proposer quelques raccourcis, notamment :

* Lundi à Vendredi
* Lundi à Samedi
* Lundi à Dimanche

**Update**: Sinon on pourrait aussi supporter utiliser <kbd>Shift</kbd> pour selectionner plusieurs jours en un clique, mais les raccourcis visibles seraient plus facils à découvrir (et marchent sur mobile !).

### PR DESCRIPTION
**Pull request by caugner**
Fixes #115.

### CODE REVIEW

**Comment by caugner** [comment]
**Note**: I still didn't manage to setup a dev environment, so please test before merging.

**Comment by caugner** [comment]
Before click on "Monday to Friday":
image

After click on "Monday to Friday":
image

**Comment by caugner** [comment]
Note that *Monday to Friday* can be combined with *Saturday*.

**Comment by PanierAvide** [comment]
Thanks for your PR :+1: I don't know if that would be clearer to hide individuals checkboxes for every day from monday to friday if "monday to friday" is checked, and show them back if unchecked. @overflorian opinions on this ?

**Comment by overflorian** [comment]
Thant you caugner. Would you have stats of days covered by opening_hours tags?
If we find that more tags are covering monday-to-saturday than monday-to-friday, we should focus on this issue instead.
And if we find that less than 20% of opening_hours are about week times, we should not change the UI.
 
Regarding your proposal I don't feel confortable with this solution that does'nt seem clear to me from a user point of view.
I propose to have a slider instead, just like in google agenda for "toute la journée" that could become "week days":
image

The expected behavior would be the same: tick monday-friday boxes.

**Comment by PanierAvide** [comment]
Looking at the stats to see which days are more used.

**Comment by PanierAvide** [comment]
Results are : 1323 `opening_hours:covid19` contains `Mo-Fr` (monday-friday), 2159 contains `Mo-Sa` (monday-saturday). Top recurring values :
```

open | 9846
-- | --
off | 5966
same | 2240
restricted | 167
Mo-Sa 09:00-19:00 | 103
Mo-Sa 08:30-19:00 | 74
24/7 | 65
Mo-Fr 07:00-21:00; Sa 07:00-20:00 | 51
Mo-Sa 09:00-18:00 | 48
Mo-Sa 09:00-12:00 | 41
Mo-Sa 08:00-20:00 | 36
00:00-23:59 | 31
Mo-Su 06:00-22:00 | 31
on | 31
07:00-15:00 | 29
Mo-Sa 09:00-20:00 | 28
Mo-Sa 08:00-19:00 | 27
Mo-Sa 10:00-18:00 | 26
Tu-Sa 09:00-12:00 | 26
Mo-Fr 09:00-12:00 | 25
Th-Su 08:30-13:00 | 25
Tu-Sa 09:00-13:00 | 25
Mo-Fr 09:00-12:00,14:00-17:00 | 23
Mo-Sa 08:30-20:00 | 22
Mo-Sa 07:00-13:00 | 21
Tu-Su 07:00-13:00 | 21
Mo-Sa 07:00-14:00 | 20
Mo-Sa 08:30-19:30 | 20
Mo-Sa 09:00-19:30 | 20
Tu-Sa 08:00-13:00 | 20
Tu-Su 07:00-14:00 | 20
Mo-Sa 06:00-20:00; Su 07:00-20:00; PH off | 19
Mo-Sa 08:00-13:00 | 19
Mo-Sa 09:00-14:00 | 19
Mo-Sa 10:00-19:00 | 19
Mo-Sa 08:00-18:00 | 18
Mo-Sa 09:00-13:00 | 17
Mo-Su 07:00-13:00 | 17
Mo-Sa 08:00-14:00 | 16
Mo-Sa 08:30-18:00 | 16
Tu | 16
Mo-Sa 07:00-20:00 | 15
Mo-Sa 09:00-17:00 | 15
Tu-Su 08:00-13:00 | 15
Mo-Sa 08:00-12:00 | 14
Mo-Su 08:00-13:00 | 14
Mo 08:00-17:00; Tu-Th 08:00-16:00; Fr 08:00-15:00; PH off | 13
Mo-Sa 08:30-18:30 | 13
Mo-Sa 09:30-18:00 | 13
```

**Comment by overflorian** [comment]
Thanks.
We already covered "open", "off" and "same" properly.
We could consider to have a monday-saturday option, why not. Unfortunately we don't have a proper mention like "week days" ("jours de semaine" ou "en semaine") for this. Therefore I wouldn't recommend it.

To be perfectly honest I think we should steal from Google Maps UI once again. I just made a screencast https://youtu.be/DVrHEWsWGy0

**Comment by caugner** [comment]
> Thanks for your PR 👍 I don't know if that would be clearer to hide individuals checkboxes for every day from monday to friday if "monday to friday" is checked, and show them back if unchecked. @overflorian opinions on this ?

This would be a regression in my point of view, because I can also use "Monday to Friday" as a shortcut for "Monday to Friday except Wednesday", just untick "Wednesday" afterwards.

You could also add "Monday to Saturday" as I had originally suggested in my issue, but the user can quickly discover that they can user "Monday to Friday" and check Saturday afterwards: Two clicks instead of six. 

I would advise against using a more sophisticated UI like a slider, because the site should also be usable by unexperienced users.

**Comment by caugner** [comment]
> Thanks.
> We already covered "open", "off" and "same" properly.
> We could consider to have a monday-saturday option, why not. Unfortunately we don't have a proper mention like "week days" ("jours de semaine" ou "en semaine") for this. Therefore I wouldn't recommend it.
> 
> To be perfectly honest I think we should steal from Google Maps UI once again. I just made a screencast https://youtu.be/7F-w29vpGBA

The video doesn't work for me, but I guess "sliding" is always more complicated than a simple tick that covers most cases (plus Monday to Saturday with an extra tick, these are checkboxes, not radios).

**Comment by caugner** [comment]
PS: Layout-wise, we could make the dialog two columns, move them right and add Monday to Saturday, Monday to Sunday and Saturday to Sunday, if that would help?

**Comment by overflorian** [comment]
I updated the video, it should work now. Please have a look. There isn't any slider: the multi-day option doesn't exist. But if you fill the time for monday and then tick tuesday, the times are automatically copied to this day. I love this solution.

**Comment by caugner** [comment]
> I updated the video, it should work now. Please have a look. There isn't any slider: the multi-day option doesn't exist. But if you fill the time for monday and then tick tuesday, the times are automatically copied to this day. I love this solution.

Hm, I don't see how this approach reduces the clicks required to select Monday to Friday.

My PR simply adds a shortcut, not more and not less, so I find it a bit irritating that you seem to argument *against* it with a UI that we don't have and that doesn't solve the underlying issue. 😕 

I'm not sure what you mean by that this isn't clear to the user. If they tick Monday to Friday individually, they will see that the shortcut gets checked, even if they add Saturday. Next time they will use it and save time. 👍

**Comment by PanierAvide** [comment]
I'm no UI expert, so I will not find any solution to this, just exposing my feeling about this whole day selecting interface (not specific to your PR) :
* Initial problem : too many checkboxes (8), UI doesn't look great, select several days is quite long
* PR Solution : adding another checkbox which is a shortcut, solves the problem but UI doesn't _look_ great
* Maybe we could distinguish more things being shortcuts and things being single days ?

**Comment by caugner** [comment]
> Initial problem : too many checkboxes (8), UI doesn't look great, select several days is quite long

For the record, this (i.e. the number of checkboxes) is *not* an issue IMO. After all, there are 7 days in a week and we cannot change this, so we need at least 7 selectable items. A checkbox is the simplest solution and any other UI like a slider would be more complex (and not necessarily reduce the number of clicks, especially not for unexperienced users).

The *real* issue is that you need to tick 5 or 6 checkboxes to select the two most common day ranges. My solution is to add one simple checkbox that reduces the number of clicks by 4. That's 80% or 67% less clicks respectively.

> Note that I hope by this discussion/debate we find the best solution for our end users, so it's worth taking time to think about this.

This is IMO a problematic approach. One person proposes a simple improvement (that works on it's own). Two other persons join the conversation to find the "best solution".

Why not go with a good solution before finding a "best solution"? You have iteratively improved the UI over the last weeks, so I really don't understand why you treat this simple and effective improvement so differently. 🤷

**Comment by caugner** [comment]
PS: For the record, I have opened #115 **two weeks ago**. The only response I got was from @overflorian who proposed UI for a special case (where `opening_hours` exists). 👍

**Comment by caugner** [comment]
How about we only show that shortcut when the country is `AT`, `CH` or `DE`?

**Comment by francois2metz** [comment]
Could you add some tests to this PR please?

**Comment by caugner** [comment]
> Could you add some tests to this PR please?

If you intend to merge the PR afterwards, sure.

**Comment by PanierAvide** [comment]
So to make some points clear before going further on discussion :
* We missed opportunity to discuss on #115 and offer a good proposal on a reasonable time. However, this project is full volunteer, so we can't expect people to answer all issues + implement code + validate PRs + translate... in a few days. We have limited time, other priorities, so it's _a missed opportunity_ but I'm personally _not sorry_ for taking time for my family and other activities.
* We, as "Ça reste ouvert" core team, are willing to work with community. As now, all PRs were accepted because after comments, we always found a solution convincing for everyone. Still, if we find some proposal not convincing, we keep the right to refuse it. However, we want to limit this as most as possible, and keep debates open and always with factual arguments.
* So, if we want to take time to discuss some subject, it's not to be annoying. It's to offer a good balance between end-users, user experience, code quality... This should be clear for everyone in order to work properly as a community.

So, after all these points, here are some suggestions on UI. We can have a checkbox hierarchy, so it has more sense that if check mon-sat it will change days to monday to saturday :


Or shortcuts can use a Material Design switch, which is more intuitive that it has different impact on form.


These two solutions are both offering a limited amount of click @caugner looks for, and a more convincing UI I and @overflorian look for. Would one of these solutions be OK for everyone ?

**Comment by caugner** [comment]
> So to make some points clear

Thank you for that, but just to be clear from my side: My criticism was about two aspects:
1. The *specific* nature of Florian's reply in that he proposes something else instead of responding to my original proposal. In this issue, he quite similarly discarded my proposal as not "clear (...) from a user point of view".
2. The apparent double standard between team contributions (move fast and iterate) and community contributions in this specific case (find "best solution").

> Or shortcuts can use a Material Design switch

Now that you say that, it seems that we had a misunderstanding here, because Florian talked about a "slider", but seemed to mean "switches" (and I didn't see any slider in his screenshot/screencast):

Sliders: https://material.io/components/sliders
Switches: https://material.io/develop/web/components/input-controls/switches/

File: components/opening_hours_editor.vue [hunk 1]
```diff
@@ -37,6 +37,16 @@
       <v-card class="pa-2">
         <template v-if="!selectTime">
           <h3 class="subtitle-1">{{ $t('opening_hours_editor.select_days') }}</h3>
+          <v-checkbox
```
Comment (caugner) [inline components/opening_hours_editor.vue:40]: Use [switch](https://material.io/develop/web/components/input-controls/switches/) for shortcuts:

```suggestion
          <v-switch
```

File: components/opening_hours_editor.vue [hunk 2]
```diff
@@ -37,6 +37,16 @@
...
+            v-for="(days, shortcut) in shortcutDays"
+            :key="shortcut"
+            :label="$t(`day_shortcuts.${shortcut}`)"
+            :value="days.every(day => selectedWeekDays.includes(day))"
+            :disabled="days.some(day => disabledWeekDays.includes(day))"
+            @click="() => toggleSelectedWeekDays(...days)"
```
Comment (caugner) [inline components/opening_hours_editor.vue:46]:
```suggestion
            @change="() => toggleSelectedWeekDays(...days)"
```

File: components/opening_hours_editor.vue [hunk 3]
```diff
@@ -151,6 +164,15 @@ export default {
...
       this.dialog = false;
+    },
+
+    toggleSelectedWeekDays(...days) {
+      const allSelected = days.every(day => this.selectedWeekDays.includes(day));
+      if (allSelected) {
+        this.selectedWeekDays = this.selectedWeekDays.filter(day => !days.includes(day));
+      } else {
+        this.selectedWeekDays = this.weekDays.filter(day => this.selectedWeekDays.includes(day) || days.includes(day));
+      }
```
Comment (caugner) [inline components/opening_hours_editor.vue:175]:
```suggestion
    toggleSelectedWeekDays(value ...days) {
      if (value) {
        this.selectedWeekDays = this.weekDays.filter(day => this.selectedWeekDays.includes(day) || days.includes(day));
      } else {
        this.selectedWeekDays = this.selectedWeekDays.filter(day => !days.includes(day));
      }
```

**Comment by caugner** [comment]
@overflorian @PanierAvide 

weekday-shortcut-switch

**Comment by PanierAvide** [comment]
Looks great ! Can we switch to monday-saturday as it is the most common range of days in `opening_hours:covid19` values ?

File: components/opening_hours_editor.vue [hunk 4]
```diff
@@ -78,6 +88,9 @@ export default {
 
   data() {
     return {
+      shortcutDays: {
+        'weekdays': ['mo', 'tu', 'we', 'th', 'fr']
```
Comment (caugner) [inline components/opening_hours_editor.vue:92]:
```suggestion
        'weekdays': ['mo', 'tu', 'we', 'th', 'fr', 'sa']
```

File: locales/en.json [hunk 5]
```diff
@@ -35,6 +35,9 @@
     "sunday": "Sunday",
     "ph": "Public holidays"
   },
+  "day_shortcuts": {
+    "weekdays": "Monday to Friday"
```
Comment (caugner) [inline locales/en.json:39]:
```suggestion
    "weekdays": "Monday to Saturday"
```

**Comment by caugner** [comment]
weekday-shortcut-switch

**Comment by overflorian** [comment]
Looks good to me.
Sorry for the misunderstanding sliders/switches @caugner

**Comment by PanierAvide** [comment]
Merged, thanks ! :+1:

**Comment by caugner** [comment]
Thanks for your merging @PanierAvide! 👍 

And sorry for the misunderstanding from my part as well, and for my tone. I should have read your original comments more carefully @overflorian.

**Comment by overflorian** [comment]
That's all right, no worries. thanks for your commitment. We took the time to discuss and produce the best version of the feature all together :)

**Pull request merged**